From be7c406f425eb08493d8f595774f340b1fa137cf Mon Sep 17 00:00:00 2001
From: Brett Mastbergen <bmastbergen@ciq.com>
Date: Tue, 25 Nov 2025 12:37:01 -0500
Subject: [PATCH] cifs: fix lease break oops in xfstest generic/098

jira VULN-131073
cve-pre CVE-2025-38527
commit-author Steve French <stfrench@microsoft.com>
commit c774e6779f38bf36f0cce65e30793704bab4b0d7
upstream-diff Applied to fs/cifs/file.c since it hasn't been
              moved to fs/smb/client/file.c in this kernel

umount can race with lease break so need to check if
tcon->ses->server is still valid to send the lease
break response.

	Reviewed-by: Bharath SM <bharathsm@microsoft.com>
	Reviewed-by: Shyam Prasad N <sprasad@microsoft.com>
Fixes: 59a556aebc43 ("SMB3: drop reference to cfile before sending oplock break")
	Signed-off-by: Steve French <stfrench@microsoft.com>
(cherry picked from commit c774e6779f38bf36f0cce65e30793704bab4b0d7)
	Signed-off-by: Brett Mastbergen <bmastbergen@ciq.com>
---
 fs/cifs/file.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/fs/cifs/file.c b/fs/cifs/file.c
index 649398d988ce..6313b352853a 100644
--- a/fs/cifs/file.c
+++ b/fs/cifs/file.c
@@ -5112,9 +5112,13 @@ void cifs_oplock_break(struct work_struct *work)
 	 * disconnected since oplock already released by the server
 	 */
 	if (!oplock_break_cancelled) {
-		rc = tcon->ses->server->ops->oplock_response(tcon, persistent_fid,
+		/* check for server null since can race with kill_sb calling tree disconnect */
+		if (tcon->ses && tcon->ses->server) {
+			rc = tcon->ses->server->ops->oplock_response(tcon, persistent_fid,
 				volatile_fid, net_fid, cinode);
-		cifs_dbg(FYI, "Oplock release rc = %d\n", rc);
+			cifs_dbg(FYI, "Oplock release rc = %d\n", rc);
+		} else
+			pr_warn_once("lease break not sent for unmounted share\n");
 	}
 
 	cifs_done_oplock_break(cinode);
-- 
2.47.3

